torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+
+_spmd_lib_impl = torch.library.Library("_spmd", "IMPL")
+for dispatch_key in ("CPU", "CUDA", "Meta"):
+    _spmd_lib_impl.impl("tag_grad", lambda x: x, dispatch_key)
+
+
+class DataParallelStyle(Enum):
+    """
+    We have three types of Data Parallel style:
+    1. DEFAULT: the default data parallel style, which is to represent a mixed
+                replicate and fully shard behavior. For each parameter that is able
+                to be sharded evenly, we shard it, otherwise we would replicate the
+                parameter. This style avoids potential padding if the parameters
+                cannot be sharded evenly, but it would generate a mixed of all_reduce
+                and reduce_scatter.
+    2. REPLICATE: the data parallel style that replicates all model parameters.
+                  This is similar to the behavior of DistributedDataParallel.
+    3. FULLY_SHARD: the data parallel style that shards all model parameters. This
+                    is similar to the behavior of FullyShardedDataParallel, the
+                    difference is that FullyShardedDataParallel (ZERO-3), which
+                    shards the model using FlatParameter based sharding,
+                    while this style shard each parameter into DTensor.
+    """
+
+    DEFAULT = 0
+    REPLICATE = 1
+    FULLY_SHARD = 2
+
+
+class NodeType(Enum):
+    """
+    NodeType is a enum that records the type of the tensors in the graph.
+    This is used to determine the data parallel strategy.
+    """
+
+    PARAM = 0
+    ACT = 1
+    GRAD = 2
+    STATE = 3
+    NON_TENSOR = 4  # NON_TENSOR is to tag non tensor node (i.e. graph output)
+
+
+@dataclass
+class DataParallelStrategy(StrategyList):
+    """
+    DataParallelStrategy is a special case of StrategyList that only records
+    the "data parallel style" placement strategy for each fx Node.
+    """
+
+    node_type: NodeType
+    reduction_over_batch: bool
+
+    def __init__(
+        self,
+        node_type: NodeType,
+        startegy_list: List[PlacementStrategy],
+        reduction_over_batch: bool = False,
+    ):
+        super().__init__(startegy_list)
+        self.node_type = node_type
+        self.reduction_over_batch = reduction_over_batch
+
+    def __str__(self) -> str:
+        return f"type: {self.node_type}, {super().__str__()}"
+
+
+@contextmanager
+def gradients_tagging(params: Dict[str, torch.Tensor]):
+    """
+    This is a helper function that tags the gradient of the parameters
+    with a special tag, so that we can identify them during SPMD expansion.
+
+    It's safe to trace those hooks and we would remove those nodes later.
+    """
+
+    tagging_hooks = []
+    try:
+        for p in params.values():
+            h = p.register_hook(lambda grad: torch.ops._spmd.tag_grad(grad))
+            tagging_hooks.append(h)
+        yield
+    finally:
+        # remove those hooks after tracing
+        for h in tagging_hooks:
+            h.remove()
```

Comment:
No change needed now, but this doesn't feel like it should be necessary, it should be "obvious" at trace time which nodes are grads versus not.

Reply:
OK, so I think I initially misunderstood the function of this. You are only adding these to the final gradient for every parameter, rather than every grad value flowing through the graph. And the reason for this is that you're tracing a full train-step graph, which means that the gradients aren't actually outputs of your graph, they immediately get fed to the optimizer.